Mathlib/Algebra/Order/Antidiag/Finsupp.lean

```diff
@@ -1,167 +1,82 @@
+finset of all finitely supported functions `f : ι →₀ μ` with support contained in `s` and such that
+the sum of its values equals `n : μ`.
 
-This Finset could be viewed inside `ι → μ`,
-but the `Finsupp` condition provides a natural `DecidableEq` instance.
-
-## Main definitions
-
-* `Finset.finsuppAntidiag s n` is the finite set of all functions `f : ι →₀ μ`
-  with finite support contained in `s` and such that the sum of its values equals `n : μ`
-  That condition is expressed by `Finset.mem_finsuppAntidiag`
-* `Finset.mem_finsuppAntidiag'` rewrites the `Finsupp.sum` condition as a `Finset.sum`.
+We define it using `Finset.piAntidiag s n`, the corresponding antidiagonal in `ι → μ`.
 -/
 
-namespace Finset
+open Finsupp Function
 
 variable {ι μ μ' : Type*}
 
-open Function Finsupp
-
-section Fin
-
-variable [AddCommMonoid μ] [DecidableEq μ] [HasAntidiagonal μ]
-
-/-- `finAntidiagonal₀ d n` is the type of d-tuples with sum `n` -/
-def finAntidiagonal₀ (d : ℕ) (n : μ) : Finset (Fin d →₀ μ) :=
-  (finAntidiagonal d n).map
-    { toFun := fun f =>
-        -- this is `Finsupp.onFinset`, but computable
-        { toFun := f, support := univ.filter (f · ≠ 0), mem_support_toFun := fun x => by simp }
-      inj' := fun _ _ h => DFunLike.coe_fn_eq.mpr h }
-
-lemma mem_finAntidiagonal₀' (d : ℕ) (n : μ) (f : Fin d →₀ μ) :
-    f ∈ finAntidiagonal₀ d n ↔ ∑ i, f i = n := by
-  simp only [finAntidiagonal₀, mem_map, Embedding.coeFn_mk, ← mem_finAntidiagonal,
-    ← DFunLike.coe_injective.eq_iff, Finsupp.coe_mk, exists_eq_right]
+namespace Finset
+section AddCommMonoid
+variable [DecidableEq ι] [AddCommMonoid μ] [HasAntidiagonal μ] [DecidableEq μ] {s : Finset ι}
+  {n : μ} {f : ι →₀ μ}
 
-lemma mem_finAntidiagonal₀ (d : ℕ) (n : μ) (f : Fin d →₀ μ) :
-    f ∈ finAntidiagonal₀ d n ↔ sum f (fun _ x => x) = n := by
-  rw [mem_finAntidiagonal₀', sum_of_support_subset f (subset_univ _) _ (fun _ _ => rfl)]
+/-- The finset of functions `ι →₀ μ` with support contained in `s` and sum `n`. -/
+def finsuppAntidiag (s : Finset ι) (n : μ) : Finset (ι →₀ μ) :=
+  (piAntidiag s n).attach.map ⟨fun f ↦ ⟨s.filter (f.1 · ≠ 0), f.1, by
+    simpa using (mem_piAntidiag.1 f.2).2⟩, fun f g hfg ↦ Subtype.ext (congr_arg (⇑) hfg)⟩
 
-end Fin
+@[simp] lemma mem_finsuppAntidiag : f ∈ finsuppAntidiag s n ↔ s.sum f = n ∧ f.support ⊆ s := by
+  simp [finsuppAntidiag, ← DFunLike.coe_fn_eq, subset_iff]
 
-section finsuppAntidiag
+lemma mem_finsuppAntidiag' :
+    f ∈ finsuppAntidiag s n ↔ f.sum (fun _ x ↦ x) = n ∧ f.support ⊆ s := by
+  simp only [mem_finsuppAntidiag, and_congr_left_iff]
+  rintro hf
+  rw [sum_of_support_subset (N := μ) f hf (fun _ x ↦ x) fun _ _ ↦ rfl]
 
-variable [DecidableEq ι]
-variable [AddCommMonoid μ] [HasAntidiagonal μ] [DecidableEq μ]
+@[simp] lemma finsuppAntidiag_empty_zero : finsuppAntidiag (∅ : Finset ι) (0 : μ) = {0} := by
+  ext f; simp [finsuppAntidiag, ← DFunLike.coe_fn_eq (g := f), -mem_piAntidiag, eq_comm]
```

Comment:
```suggestion
  ext f; simp [finsuppAntidiag, ← DFunLike.coe_fn_eq (g := f), eq_comm]
```
Or is a precise reason to explicitely remove `mem_piAntidiag`?

Reply:
Oh, it's not needed anymore. Well spotted